docs/settings.rst
.. _settings:

Settings
========

Settings reference
~~~~~~~~~~~~~~~~~~


Here is a complete list of settings which control default PynamoDB behavior.

connect_timeout_seconds
-----------------------

Default: ``60``

The time till a timeout exception is thrown when attempting to make a connection.


read_timeout_seconds
-----------------------

Default: ``60``

The time till a timeout exception is thrown when attempting to read from a connection.


max_retry_attempts
------------------

Default: ``3``

The number of times to retry certain failed DynamoDB API calls. The most common cases eligible for
retries include ``ProvisionedThroughputExceededException`` and ``5xx`` errors.


base_backoff_ms
---------------

Default: ``25``

The base number of milliseconds used for `exponential backoff and jitter
<https://www.awsarchitectureblog.com/2015/03/backoff.html>`_ on retries.


region
------

Default: ``"us-east-1"``

The default AWS region to connect to.


max_pool_connections
--------------------

Default: ``10``

The maximum number of connections to keep in a connection pool.


extra_headers
--------------------

Default: ``None``

A dictionary of headers that should be added to every request. This is only useful
when interfacing with DynamoDB through a proxy, where headers are stripped by the
proxy before forwarding along. Failure to strip these headers before sending to AWS
will result in an ``InvalidSignatureException`` due to request signing.


allow_rate_limited_scan_without_consumed_capacity
-------------------------------------------------

Default: ``False``

If ``True``, ``rate_limited_scan()`` will proceed silently (without
rate limiting) if the DynamoDB server does not return consumed
capacity information in responses. If ``False``, scans will fail
should the server not return consumed capacity information in an
effort to prevent unintentional capacity usage..

Overriding settings
~~~~~~~~~~~~~~~~~~~

Default settings may be overridden by providing a Python module which exports the desired new values.
Set the ``PYNAMODB_CONFIG`` environment variable to an absolute path to this module or write it to
``/etc/pynamodb/global_default_settings.py`` to have it automatically discovered.


docs/release_notes.rst
Release Notes
=============

v4.0.0a1
--------

:date: 2019-04-10

This is major release and contains breaking changes. Please read the notes below carefully.

Given that `botocore` has moved to using `urllib3` directly for making HTTP requests, we'll be doing the same (via `botocore`). This means the following:

 * The `session_cls` option is no longer supported. <TODO: discuss replacement opts>
 * The `request_timeout_seconds` parameter is no longer supported. `connect_timeout_seconds` and `read_timeout_seconds` are now available instead.

Other changes in this release:

 * Fixes moto support through implementing the botocore "before-send" hook. Other botocore hooks remain unimplemented.


v3.3.3
------

:date: 2019-01-15

This is a backwards compatible, minor release.

Fixes in this release:

 * Legacy boolean attribute migration fix. (#538)
 * Correctly package type stubs. (#585)

Contributors to this release:

 * @vo-va


v3.3.2
------

:date: 2019-01-03

This is a backwards compatible, minor release.

Changes in this release:

 * Built-in support for mypy type stubs, superseding those in python/typeshed. (#537)


v3.3.1
------

:date: 2018-08-30

This is a backwards compatible, minor bug fix release.

Fixes in this release:

* Clearer error message on missing consumed capacity during rate-limited scan. (#506)
* Python 3 compatibility in PageIterator. (#535)
* Proxy configuration changes in botocore>=1.11.0. (#531)

Contributors to this release:

* @ikonst
* @zetaben
* @ningirsu


v3.3.0
------

:date: 2018-05-09

This is a backwards compatible, major bug fix release.

New features in this release:


* Support scan operations on secondary indexes. (#141, #392)
* Support projections in model get function. (#337, #403)
* Handle values from keys when batch get returns unprocessed keys. (#252, #376)
* Externalizes AWS Credentials. (#426)
* Add migration support for LegacyBooleanAttribute. (#404, #405)
* Rate limited Page Iterator. (#481)


Fixes in this release:

* Thread-safe client creation in botocore. (#153, #393)
* Use attr.get_value(value) when deserialize. (#450) 
* Skip null attributes post serialization for maps. (#455)
* Fix deserialization bug in BinaryAttribute and BinarySetAttribute. (#459, #480)
* Allow MapAttribute instances to be used as the RHS in expressions. (#488)
* Return the correct last_evaluated_key for limited queries/scans. (#406, #410)
* Fix exclusive_start_key getting lost in PageIterator. (#421)
* Add python 3.5 for Travis ci builds. (#437)

Contributors to this release:

* @jpinner-lyft
* @scode
* @behos
* @jmphilli
* @drewisme
* @nicysneiros
* @jcomo
* @kevgliss
* @asottile 
* @harleyk
* @betamoo


v3.2.1
------

:date: 2017-10-25

This is a backwards compatible, minor bug fix release.

Removed features in this release:

* Remove experimental Throttle api. (#378)

Fixes in this release:

* Handle attributes that cannot be retrieved by getattr. Fixes #104 (#385)
* Model.refresh() should reset all model attribuets. Fixes #166 (#388)
* Model.loads() should deserialize using custom attribute names. Fixes #168 (#387)
* Deserialize hash key during table loads. Fixes #143 (#386)
* Support pagination in high-level api query and scan methods. Fixes #50, #118, #207, and #248 (#379)
* Don't serialize null nested attributed. Fixes #240 and #309 (#375)
* Legacy update item subset removal using DELETE operator. Fixes #132 (#374)

Contributors to this release:

* @jpinner-lyft


v3.2.0
------

:date: 2017-10-13

This is a backwards compatible, minor release.

This release updates PynamoDB to interact with Dynamo via the current version of Dynamo's API.
Condition and update expressions can now be created from attributes and used in model operations.
Legacy filter and attribute update keyword arguments have been deprecated. Using these arguments
will cause a warning to be logged.

New features in this release:

* Add support for current version of `DynamoDB API <http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Appendix.CurrentAPI.html>`_
* Improved ``MapAttribute`` item assignment and access.

Contributors to this release:

* @jpinner-lyft


v3.2.0rc2
---------

:date: 2017-10-09

This is a backwards compatible, release candidate.

This release candidate allows dereferencing raw ``MapAttributes`` in condition expressions.
It also improves ``MapAttribute`` assignment and access.

Contributors to this release:

* @jpinner-lyft


v3.2.0rc1
---------

:date: 2017-09-22

This is a backwards compatible, release candidate.

This release candidate updates PynamoDB to interact with Dynamo via the current version of Dynamo's API. 
It deprecates some internal methods that were used to interact with Dynamo that are no longer relevant. 
If your project was calling those low level methods a warning will be logged.

New features in this release:

* Add support for current version of `DynamoDB API <http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Appendix.CurrentAPI.html>`_

Contributors to this release:

* @jpinner-lyft


v3.1.0
------

:date: 2017-07-07

This is a backwards compatible, minor release.

Note that we now require ``botocore>=1.2.0``; this is required to support the 
``consistent_read`` parameter when scanning.

Calling ``Model.count()`` without a ``hash_key`` and *with* ``filters`` will
raise a ``ValueError``, as it was previously returning incorrect results.

New features in this release:

* Add support for signals via blinker (#278)

Fixes in this release:

* Pass batch parameters down to boto/dynamo (#308)
* Raise a ValueError if count() is invoked with no hash key AND filters (#313)
* Add consistent_read parameter to Model.scan (#311)

Contributors to this release:

* @jmphilli
* @Lordnibbler
* @lita


v3.0.1
------

:date: 2017-06-09

This is a major release with breaking changes.

``MapAttribute`` now allows pythonic access when recursively defined.

pynamodb/settings.py
import imp
import logging
import os
import warnings
from os import getenv

from botocore.vendored import requests

log = logging.getLogger(__name__)

default_settings_dict = {
    'connect_timeout_seconds': 60,
    'read_timeout_seconds': 60,
    'max_retry_attempts': 3,
    'base_backoff_ms': 25,
    'region': 'us-east-1',
    'max_pool_connections': 10,
    'extra_headers': None,
    'allow_rate_limited_scan_without_consumed_capacity': False,
}

OVERRIDE_SETTINGS_PATH = getenv('PYNAMODB_CONFIG', '/etc/pynamodb/global_default_settings.py')

override_settings = {}
if os.path.isfile(OVERRIDE_SETTINGS_PATH):
    override_settings = imp.load_source('__pynamodb_override_settings__', OVERRIDE_SETTINGS_PATH)
    if hasattr(override_settings, 'session_cls') or hasattr(override_settings, 'request_timeout_seconds'):
        warnings.warn("The `session_cls` and `request_timeout_second` options are no longer supported")
    log.info('Override settings for pynamo available {0}'.format(OVERRIDE_SETTINGS_PATH))
else:
    log.info('Override settings for pynamo not available {0}'.format(OVERRIDE_SETTINGS_PATH))
    log.info('Using Default settings value')


def get_settings_value(key):
    """
    Fetches the value from the override file.
    If the value is not present, then tries to fetch the values from constants.py
    """
    if hasattr(override_settings, key):
        return getattr(override_settings, key)

    if key in default_settings_dict:
        return default_settings_dict[key]

    return None
